weave/trace_server/calls_query_builder/optimization_builder.py

```diff
@@ -0,0 +1,448 @@
+        return None
+
+    def process_in(self, operation: tsi_query.InOperation) -> Optional[str]:
+        return None
+
+    def process_gt(self, operation: tsi_query.GtOperation) -> Optional[str]:
+        return None
+
+    def process_gte(self, operation: tsi_query.GteOperation) -> Optional[str]:
+        return None
+
+    def finalize_sql(self, result: Optional[str]) -> Optional[str]:
+        """Final step to make valid SQL for the calls query."""
+        if result:
+            return f"AND {result}"
+        return None
+
+
+class StringOptimizationProcessor(QueryOptimizationProcessor):
+    def process_eq(self, operation: tsi_query.EqOperation) -> Optional[str]:
+        return _create_like_optimized_eq_condition(operation, self.pb, self.table_alias)
+
+    def process_contains(self, operation: tsi_query.ContainsOperation) -> Optional[str]:
+        return _create_like_optimized_contains_condition(
+            operation, self.pb, self.table_alias
+        )
+
+    def process_in(self, operation: tsi_query.InOperation) -> Optional[str]:
+        return _create_like_optimized_in_condition(operation, self.pb, self.table_alias)
+
+
+class IdOptimizationProcessor(QueryOptimizationProcessor):
+    def process_gt(self, operation: tsi_query.GtOperation) -> Optional[str]:
+        return _create_datetime_optimization_sql(operation, self.pb, self.table_alias)
+
+    def process_gte(self, operation: tsi_query.GteOperation) -> Optional[str]:
+        return _create_datetime_optimization_sql(operation, self.pb, self.table_alias)
+
+    def finalize_sql(self, result: Optional[str]) -> Optional[str]:
+        if result:
+            with_otel_id_filter = _add_otel_id_filter(self.table_alias, result)
+            return super().finalize_sql(with_otel_id_filter)
+        return None
+
+
+def apply_processor(
+    processor: QueryOptimizationProcessor, operation: tsi_query.Operation
+) -> Optional[str]:
+    if isinstance(operation, tsi_query.AndOperation):
+        return processor.process_and(operation)
+    elif isinstance(operation, tsi_query.OrOperation):
+        return processor.process_or(operation)
+    elif isinstance(operation, tsi_query.NotOperation):
+        return processor.process_not(operation)
+    elif isinstance(operation, tsi_query.EqOperation):
+        return processor.process_eq(operation)
+    elif isinstance(operation, tsi_query.ContainsOperation):
+        return processor.process_contains(operation)
+    elif isinstance(operation, tsi_query.InOperation):
+        return processor.process_in(operation)
+    elif isinstance(operation, tsi_query.GtOperation):
+        return processor.process_gt(operation)
+    elif isinstance(operation, tsi_query.GteOperation):
+        return processor.process_gte(operation)
+    return None
+
+
+class OptimizationConditions(BaseModel):
+    str_filter_opt_sql: Optional[str] = None
+    id_datetime_filters_sql: Optional[str] = None
```

Comment:
could correlate these with the classes. Like one handler that registers the class and calls finalize